File: LowestCommonAncestor.java
package sweng;
public class LowestCommonAncestor {
	
	public Node head;
	
	//Node class to represent Nodes of the binary tree with left, right and parent nodes as well as a key represented by a character
	public class Node{  
		private char key;
		public Node left, right, parent;
		private int value;
	}

	public void createHead(char key,int value){
		Node head = new Node();
		head.key = key;
		head.left = null;
		head.right = null;
		head.value = value;
		this.head = head;
	}
	
	public char returnKey(Node node){
		return node.key;
	}
	
	public int returnValue(Node node){
		return node.value;
	}

	public void addNode(char key, int value){
		Node newNode = new Node();
		newNode.key = key;
		newNode.left = null;
		newNode.right = null;	
		newNode.value = value;
		addNode(newNode, head);
	}
	
	private void addNode(Node newNode, Node head){
		if(newNode.key < head.key){
			if(head.left == null){
				head.left = newNode;
				newNode.parent = head;
			}
			else{
				addNode(newNode, head.left);
			}
		}
		else{
			if(head.right == null){
				head.right = newNode;
				newNode.parent = head;
			}
			else{
				addNode(newNode, head.right);
			}
		}
	}
	
	public Node findNode(char key){
		if(key == head.key){
			return head;
		}
		else {
			return findNode(key, head);
		}
	}
	private Node findNode(char key, Node head){
		Node ret = null;
		if(head.left != null){
			if (head.left.key == key){
				return head.left;
			}
			else{
				ret = findNode(key, head.left);
				if(ret != null){
					return ret;
				}
			}
		}
		if(head.right != null){
			if(head.right.key == key){
				return head.right;
			}
			else{
				ret = findNode(key, head.right);
				if(ret != null){
					return ret;
				}
			}
		}
		return null;
	}	
	public int depth(char key){
		int depth = 0;
		Node node = findNode(key);
		while(node.parent != null){
			depth++;
			node = node.parent;
		}
		return depth;
	}

	public Node lowestCommonAncestor(Node root, Node p, Node q) {
	    if(root==null)
	        return null;
	    if(root==p || root==q)
	        return root;
	    Node l = lowestCommonAncestor(root.left, p, q);
	    Node r = lowestCommonAncestor(root.right, p, q);
	    
	    if(l!=null&&r!=null){
	        return root;
	    }else if(l==null&&r==null){
	        return null;
	    }else{
	        return l==null?r:l;
	    }
	}
	
	public String prettyPrintKeys() {
    	if (head == null){
    		return "-null\n";
    	}
    	else if (head.left == null && head.right == null){
    		return "-" + head.key + "\n" + " |-null\n" + "  -null\n";
    	}
    	return prettyPrintKeys(head,"");
    }
    
    private String prettyPrintKeys(Node x,String prefix){
    	if (x == null){
    		return prefix + "-null\n";
    	}
    	if (x.left != null){
    		return prefix + "-" + x.key + "\n" + prettyPrintKeys(x.left, prefix + " |") + prettyPrintKeys(x.right, prefix + "  ");
    	}
    	else if (x.left == null){
    		return prefix + "-" + x.key + "\n" + prefix + " |" + "-null\n" + prettyPrintKeys(x.right,prefix + "  ");
    	}
    	return null;
    }
	
	

	public String helloWorld(){
		return "HelloWorld";
	}
	
	public static void main(String[] args) {
		LowestCommonAncestor lca = new LowestCommonAncestor();
		lca.createHead('e',0);
		lca.addNode('b', 1);
		lca.addNode('f', 1);
		lca.addNode('a', 1);
		lca.addNode('g', 1);
		String tree = lca.prettyPrintKeys();
		System.out.print(tree);
	}

}

File: LowestCommonAncestor.py
#geeksforgeeks.com

class Node:
    def __init__(self, key):
        self.key =  key
        self.left = None
        self.right = None
 

def findPath( root, path, k):
 
    # Baes Case
    if root is None:
        return None


    if root.key == path or root.key == k :
        return root

    if path == k:
        return path.key

    dag = []
    i = 0

    while(i < len(path.pred)):
        j = 0
        while(j < len(k.pred)):
            if(path.pred[i].key == k.pred[j].key):
                dag.append(path[i].key)
                j = j + 1
            else:
                j = j+1
        i = i + 1
    if(dag == []):
        if(path.key > k.key):
            dag.append(findPath(root, path.pred[0], k))
        else:
            dag.append(findPath(root, path, k.pred[0]))

    return max(dag)


root = Node(1)
r2 = Node(2)
r3 = Node(3)
r4 = Node(4)
r5 = Node(5)
r6 = Node(6)
root.succ = [r2, r3, r4, r5]
r2.succ = [r4]
r2.pred = [root]
r3.succ = [r4, r5]
r3.pred = [root]
r4.succ = [r5]
r4.pred = [r2, r3, root]
r5.pred = [r3, r4, root]
r6.pred = [r4]
